hibernate-core/src/main/java/org/hibernate/persister/entity/SingleTableEntityPersister.java

```diff
@@ -349,10 +351,10 @@ public SingleTableEntityPersister(
 			);
 
 			// SUBCLASSES
-			final List<Subclass> subclasses = persistentClass.getSubclasses();
+			final List<Subclass> subclasses = persistentClass.getSubclasses().stream().sorted(Comparator.comparing(PersistentClass::getEntityName)).collect(Collectors.toList());
```

Comment:
The subclasses need to be ordered by the entity name to match the TreeMap with entity names keys used for `AbstractEntityPersister.subclassMappingTypes`. Otherwise the column orders can get switched up incorrectly and return the wrong values.